File: demo-test/src/test/java/io/fries/demo/test/cucumber/wiremock/record/transformer/IdempotentStubIdTransformer.java
package io.fries.demo.test.cucumber.wiremock.record.transformer;

import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder;
import com.github.tomakehurst.wiremock.common.FileSource;
import com.github.tomakehurst.wiremock.extension.Parameters;
import com.github.tomakehurst.wiremock.extension.StubMappingTransformer;
import com.github.tomakehurst.wiremock.stubbing.StubMapping;

import java.util.HashSet;
import java.util.Set;
import java.util.UUID;
import java.util.concurrent.atomic.AtomicInteger;

import static java.nio.charset.StandardCharsets.UTF_8;

public class IdempotentStubIdTransformer extends StubMappingTransformer {

    private final String serverName;
    private final AtomicInteger requestsCounter;
    private final Set<UUID> previouslyAppliedOn;

    public IdempotentStubIdTransformer(final String serverName) {
        this.serverName = serverName;
        this.requestsCounter = new AtomicInteger();
        this.previouslyAppliedOn = new HashSet<>();
    }

    @Override
    public StubMapping transform(final StubMapping stubMapping, final FileSource files, final Parameters parameters) {
        if (previouslyAppliedOn.contains(stubMapping.getId())) {
            return stubMapping;
        }

        final var idempotentId = generateIdempotentUUID();
        updateResponseBodyFileName(stubMapping, files, idempotentId);
        stubMapping.setId(idempotentId);
        stubMapping.setName(serverName);

        previouslyAppliedOn.add(idempotentId);
        return stubMapping;
    }

    private UUID generateIdempotentUUID() {
        final var seed = String.format("%s-%d", serverName, requestsCounter.incrementAndGet()).getBytes(UTF_8);
        return UUID.nameUUIDFromBytes(seed);
    }

    private void updateResponseBodyFileName(final StubMapping stubMapping, final FileSource files, final UUID idempotentId) {
        final var response = stubMapping.getResponse();
        final var bodyFileName = response.getBodyFileName();
        if (bodyFileName == null) {
            return;
        }

        final var idempotentFileName = bodyFileName
                .replace(stubMapping.getName(), serverName)
                .replace(stubMapping.getId().toString(), idempotentId.toString());
        final var responseBody = files.getTextFileNamed(bodyFileName).readContentsAsString();
        files.deleteFile(bodyFileName);
        files.writeTextFile(idempotentFileName, responseBody);

        final var updatedResponse = ResponseDefinitionBuilder.like(response)
                .withBodyFile(idempotentFileName)
                .build();
        stubMapping.setResponse(updatedResponse);
    }

    @Override
    public String getName() {
        return "idempotent-stub-id-transformer";
    }
}

File: demo-test/src/test/kotlin/io/fries/demo/test/cucumber/wiremock/record/transformer/IdempotentStubIdTransformer.kt
package io.fries.demo.test.cucumber.wiremock.record.transformer

import com.github.tomakehurst.wiremock.client.ResponseDefinitionBuilder
import com.github.tomakehurst.wiremock.common.FileSource
import com.github.tomakehurst.wiremock.extension.Parameters
import com.github.tomakehurst.wiremock.extension.StubMappingTransformer
import com.github.tomakehurst.wiremock.stubbing.StubMapping
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.concurrent.atomic.AtomicInteger

class IdempotentStubIdTransformer(private val serverName: String) : StubMappingTransformer() {

    private val requestsCounter: AtomicInteger = AtomicInteger()
    private val previouslyAppliedOn: MutableSet<UUID> = mutableSetOf()

    override fun transform(stubMapping: StubMapping, files: FileSource, parameters: Parameters?) = stubMapping.apply {
        if (previouslyAppliedOn.contains(id)) {
            return this
        }

        val idempotentId = generateIdempotentUUID()
        updateResponseBody(files, idempotentId)
        id = idempotentId
        name = serverName

        previouslyAppliedOn.add(idempotentId)
    }

    private fun generateIdempotentUUID(): UUID = "$serverName-${requestsCounter.incrementAndGet()}"
        .toByteArray(StandardCharsets.UTF_8)
        .let(UUID::nameUUIDFromBytes)

    private fun StubMapping.updateResponseBody(files: FileSource, idempotentId: UUID) = apply {
        val bodyFileName = response.bodyFileName ?: return this
        val idempotentFileName = bodyFileName
            .replace(name, serverName)
            .replace(id.toString(), idempotentId.toString())

        val responseBody = files.getTextFileNamed(bodyFileName).readContentsAsString()
        files.deleteFile(bodyFileName)
        files.writeTextFile(idempotentFileName, responseBody)

        response = ResponseDefinitionBuilder.like(response)
            .withBodyFile(idempotentFileName)
            .build()
    }

    override fun getName(): String = "idempotent-stub-id-transformer"
}